### PR TITLE
ngraph ops implementation

### PR DESCRIPTION
**Pull request by gf712**
Can now switch between vanilla shogun and ngraph backends

### CODE REVIEW

File: src/shogun/mathematics/graph/ops/ngraph/Add.h [hunk 1]
```diff
@@ -28,6 +28,15 @@ namespace shogun {
...
 		{
 		}	
+
+		void build_implementation() final
+		{
+			// TODO
+			const auto& node1 = this->m_abstract_node->get_input_nodes()[0];
+			const auto& node2 = this->m_abstract_node->get_input_nodes()[1];
+
+			std::make_shared<ngraph::op::Add>(node1, node2);
```
Comment (gf712) [inline src/shogun/mathematics/graph/ops/ngraph/Add.h]: @vigsterkr somehow the nodes need to have access to the ngraph::Node. Which means then that the Node class in Shogun needs to be aware of the operator implementation? Or there is a NGraphNode that can be casted to?

File: src/shogun/mathematics/graph/NGraph.cpp [hunk 2]
```diff
@@ -1,62 +1,126 @@
...
+	}
+	return op_it->second();
 }
 
-void NGraph::add_input_operator(const std::shared_ptr<Node>& node)
+void NGraph::add_input_operator(const std::shared_ptr<node::Node>& node)
 {
+	
+	auto input = get_operator(node);
+	m_lookup[node] = std::static_pointer_cast<detail::ngraph::InputNGraph>(input)->build_input(node);
```
Comment (gf712) [inline src/shogun/mathematics/graph/NGraph.cpp:114]: @vigsterkr unfortunately ngraph can't infer the shape at execution time from dynamic. So need to the building of the ngraph graph at execution. So need change the top Graph abstraction..

Comment (gf712) [reply src/shogun/mathematics/graph/NGraph.cpp:114]: nvm, just have to call a different op::Parameter constructor and create dynamic tensors

**Comment by gf712** [comment]
@vigsterkr tests pass on both backends!

**Comment by vigsterkr** [comment]
great work! 💃

**Pull request merged**